### PR TITLE
Finish Standardization and Rewrite

### LINKED ISSUE #625: Drop fixed path in favor of relative when calling binaries
In order to get some more headroom for folder changes should the scripts in what extent is possible use relative calling on binaries. i.e. `git` in favor of `/local/bin/git`.

- [x] Update CONTRIBUTE.md with coding style
- [ ] Review all files in the whole repository

### LINKED ISSUE #937: Rewrite of openHABian core functions
There are many places where the existing code is in need of a good tune up to make it safer in execution and to keep to a standard more consistently. 

This will take a while but I will open a PR and begin to rewrite the core of openHABian to have much safer execution and more clear outputs.

Related #936

### PR DESCRIPTION
**Pull request by ecdye**
Fixes #937
Fixes #625

Signed-off-by: Ethan Dye <mrtops03@gmail.com>

### CODE REVIEW

**Comment by mstormi** [comment]
did you look at #625, too ?

**Comment by ecdye** [comment]
> did you look at #625, too ?

Yes, that will be resolved with this as well.

Quick question, is our Amanda setup function intended to be able to create two functioning configs that work simultaneously? Meaning could one run an AWS config while also running a local USB attached device? Is that something we support, or no?

Also, is the Raspbian package version of Amanda updated enough now to enable the use of a set of SD cards for backing up onto? It appears to be the latest released version which is `3.5.1` released in December of 2017 according to <http://www.amanda.org>.

**Comment by mstormi** [comment]
> is our Amanda setup function intended to be able to create two functioning configs that work simultaneously?

You can run multiple configs in parallel and it _might_ work or not with the current code to run it twice with different input.
But I didn't explicitly build the routine to support dual setups. create_backup_config() probably does but amanda_setup() would need a rework so stuff that needs to be done only once (like adding a backup user) is only done once.

> Also, is the Raspbian package version of Amanda updated enough now to enable the use of a set of SD cards for backing up onto?

yes IIRC SD based would work with 3.5.1. But the code is untested and I don't know if there's people who would want to use it.

**Comment by ecdye** [comment]
I think we should only support one config at once, as currently is I am understanding the logic correctly it will not work properly. If there is a demand for more than one supported at the same time we can add it at a later date.

> yes IIRC SD based would work with 3.5.1. But the code is untested and I don't know if there's people who would want to use it.

Should I just remove it then? I don't think most people really want to use several SD cards to backup one system, and I don't have the capability to test it's functionality on my own HW.

**Comment by mstormi** [comment]
> I think we should only support one config at once

Agree, but has it ever been different ? Would we need to change anything or why did you ask in the first place ?


> Should I just remove it then?

yes

**Comment by ecdye** [comment]
> Agree, but has it ever been different ? Would we need to change anything or why did you ask in the first place ?

I asked because it was unclear from the code what exactly was happening without doing careful analysis.

> yes

Will do.

File: build.bash
Comment (ThomDietrich) [inline build.bash:5]: 😅

Comment (mstormi) [reply build.bash:5]: image is intentionally only built when build.bash is changed

Comment (ecdye) [reply build.bash:5]: If nothing else, we developers are certainly a lazy creative bunch. 

> Whenever there is a hard job to be done I assign it to a lazy man; he is sure to find an easy way of doing it. - Walter Chrysler

Comment (mstormi) [reply build.bash:5]: Ok Walter, I got a job for you #1078

Comment (ThomDietrich) [reply build.bash:5]: I remember times when I would regularly remove and add the period at the end of a comment. :)

File: functions/ext-storage.bash
Comment (mstormi) [inline functions/ext-storage.bash:63]: more precisely it's unlimiting the power draw to up to 1.2A on older RPi
on newer (4, not sure of 3) it has no effect (but does not do harm either)

File: functions/find.bash
Comment (mstormi) [inline functions/find.bash:117]: FALIED

File: functions/helpers.bash
Comment (mstormi) [inline functions/helpers.bash:90]: are we checking return value in the calling routine ? just a thought

File: functions/influxdb+grafana.bash
Comment (mstormi) [inline functions/influxdb+grafana.bash:10]: Correct for the time being but should not be hard to provide a version that also does unattended install taking its values off openhabian.conf, should it ....

File: functions/influxdb+grafana.bash
Comment (mstormi) [inline functions/influxdb+grafana.bash:262]: FALIED

**Review by mstormi — CHANGES_REQUESTED** [review]
I'm reviewing in stages, am about half way into

File: build-image/first-boot.bash
Comment (ecdye) [reply build-image/first-boot.bash:68]: I personally prefer piping input as I find it safer than redirecting it. Unless there is an issue I would prefer it that way.

File: functions/backup.bash
Comment (ecdye) [reply functions/backup.bash:10]: I'll look into it. It should be pretty simple to make it work in unattended.

File: build-image/first-boot.bash
Comment (mstormi) [reply build-image/first-boot.bash:68]: I'm fine with that, just reminded me of that

File: functions/backup.bash
Comment (ecdye) [reply functions/backup.bash:109]: No, remember I talked about making the `amanda_install()` function to allow it to be called from uninteractive and then use the `create_amanda_config` inside of setup_mirrior_SD to complete the setup. look at it again.

File: functions/backup.bash
Comment (ecdye) [reply functions/backup.bash:163]: It is fine the function is designed to handle it.

File: functions/helpers.bash
Comment (ecdye) [reply functions/helpers.bash:90]: Nope just the string if I understand bash logic correctly.

File: functions/influxdb+grafana.bash
Comment (ecdye) [reply functions/influxdb+grafana.bash:10]: No, if that becomes a desired feature it should be fairly simple.

File: functions/backup.bash
Comment (mstormi) [reply functions/backup.bash:109]: So which of those question is "no" your answer on? if you rebased ?
I remember we talked about but IIRC the outcome was that I asked you to take my most recent Amanda code and eventually standardize/rewrite it again. BTW it already implements both, interactive AND unattended.
Current code calls amanda_setup() from unattended openhabian-setup (and that your PR does not change).

File: functions/ext-storage.bash
Comment (ecdye) [reply functions/ext-storage.bash:63]: Correct, I just left that comment alone and didn't update it but I will change it to reflect reality.

File: functions/helpers.bash
Comment (mstormi) [reply functions/helpers.bash:90]: ok but yes when it fails the return string _should_ be empty and that the calling routines check against so it should do.

File: functions/influxdb+grafana.bash
Comment (mstormi) [reply functions/influxdb+grafana.bash:10]: as mentioned earlier we should provide interactive AND unattended for every new install routine (and ultimately for every old one, too).
I leave it to you though when to tackle that.

File: functions/zram.bash
Comment (mstormi) [inline functions/zram.bash:46]: short form please

File: functions/zram.bash
Comment (mstormi) [inline functions/zram.bash:124]: will fail to remove ZRAM from AWS if this here already fails

File: includes/mount_template
Comment (mstormi) [inline includes/mount_template:6]: double Before= and DefaultDependencies=

File: functions/system.bash
Comment (mstormi) [inline functions/system.bash:306]: thorough 👍

File: functions/system.bash
Comment (mstormi) [inline functions/system.bash:332]: huh? does shellcheck like that ?

**Review by mstormi — CHANGES_REQUESTED** [review]
two more files to go but for today it's enough

File: functions/packages.bash
Comment (mstormi) [inline functions/packages.bash:26]: we eventually might need a mkdir here (for RuntimeDirectory?), see #1112

File: functions/packages.bash
Comment (mstormi) [inline functions/packages.bash:105]: another constant

File: functions/packages.bash
Comment (mstormi) [inline functions/packages.bash:174]: 2 more constants

File: functions/packages.bash
Comment (mstormi) [inline functions/packages.bash:541]: constant

**Review by mstormi — CHANGES_REQUESTED** [review]

File: functions/backup.bash
Comment (ecdye) [reply functions/backup.bash:109]: Yes I rebased, no it should not be called in unattended.

Yes it does, It calls it from you mirror sd setup function. I missed removing the call in openhabian-setup.sh but it has since been changed and corrected. look closely.

File: functions/zram.bash
Comment (ecdye) [reply functions/zram.bash:124]: What line are you talking about? GitHub's code review system makes it very ambiguous at times.

File: functions/system.bash
Comment (ecdye) [reply functions/system.bash:332]: It doesn't care. Otherwise I would have a lot of ShellCheck error notifications on GitHub.

File: functions/packages.bash
Comment (ecdye) [reply functions/packages.bash:26]: Do you want me just to go ahead and add it now?

Comment (mstormi) [reply functions/packages.bash:26]: yeah why not. Attention though, this is just an assumption, I'm not sure this is the reason for the error message but you'll see in the testbuild won't you

File: functions/zram.bash
Comment (mstormi) [reply functions/zram.bash:124]: L126, the if test for an AWS config. Won't reach if the previous test fails.

File: functions/backup.bash
Comment (mstormi) [reply functions/backup.bash:109]: ok, I checked again and yes it looks almost fine now.
However I would still like amanda_setup not to be called nested from setup_mirror_SD but from main script (albeit this time *after* setup_mirror_SD).
One reason is I really dislike nesting in general as it potentially fogs dependencies.
The other is you can select setup_mirror_SD from the 5X menu _without_ that you want to setup Amanda. Current PR code would do that though.

File: functions/packages.bash
Comment (ecdye) [reply functions/packages.bash:26]: yep

File: functions/zram.bash
Comment (ecdye) [reply functions/zram.bash:124]: I'll look at it, GitHub didn't even show that line in its snippet.

File: functions/backup.bash
Comment (ecdye) [reply functions/backup.bash:109]: OK, `amanda_setup` is not called in the `setup_mirror_SD` but the install function `amanda_install` is. I would like to keep it where it is because the `create_amanda_config` is called in the `setup_mirror_SD` right after `amanda_install`. The way I have it set up makes it fairly clear about dependencies and moving it outside would make it so that the config would not be created.

Being able to install SD mirroring without Amanda was intentional in an interactive setup as the option to install Amanda is a separate option and it allows more finite control over the system and what is set up. I think it would be better to simply recommend to use the leftover space to setup Amanda after mirroring setup is finished. Then Amanda can also be setup to have proper mailing support as well.

Comment (mstormi) [reply functions/backup.bash:109]: Ok then. Not my preference but I don't mind to keep it as-is.

File: functions/backup.bash
Comment (ecdye) [reply functions/backup.bash:18]: Nope because they are supposed to be declared after any quick exits. I put in the style guide.

Comment (mstormi) [reply functions/backup.bash:18]: ok

File: functions/backup.bash
Comment (ecdye) [reply functions/backup.bash:326]: Yes, but I don't really know how to do that. I'll create a separate issue to track removing cron jobs and replacing them with systemd timers.

File: functions/zram.bash
Comment (ecdye) [reply functions/zram.bash:124]: Yes it should, there are two possible config files, if the first config file does not exist there is a possibility that the other one will.

File: functions/backup.bash
Comment (ecdye) [reply functions/backup.bash]: Done

File: functions/backup.bash
Comment (ecdye) [reply functions/backup.bash]: Done

File: functions/zram.bash
Comment (ecdye) [reply functions/zram.bash]: done

File: includes/mount_template
Comment (ecdye) [reply includes/mount_template]: done

File: functions/packages.bash
Comment (ecdye) [reply functions/packages.bash]: added

**Comment by ecdye** [comment]
I think I am almost done, tomorrow I will try to do a final test of all changes on HW to confirm everything is working and then we should be pretty much ready.

File: functions/backup.bash
Comment (mstormi) [reply functions/backup.bash]: thanks. I'd suggest to additionally intro backuppath variable in openhabian.conf

File: functions/packages.bash
Comment (mstormi) [reply functions/packages.bash]: and did it fix the error?

**Review by mstormi — APPROVED** [review]
LGTM

**Comment by mstormi** [comment]
@ecdye Ethan I saw your request for maintainership and apologies I haven't answered it yet. I had forwarded this to the other maintainers but did not get any response either. Could you point me at your request please - I don't find it any more.
I'm not sure how to proceed if I don't get any response, maybe need to talk to Kai about it.

**Comment by ecdye** [comment]
> @ecdye Ethan I saw your request for maintainership and apologies I haven't answered it yet. I had forwarded this to the other maintainers but did not get any response either. Could you point me again at your request please - I don't find it any more.
> I'm not sure how to proceed if I don't get any response, maybe need to talk to Kai about it.

The one from months ago, when I first joined? I don't recall making one since.

**Comment by mstormi** [comment]
> The one from months ago, when I first joined? I don't recall making one since.

No a later one also nominating Holger, something like 6 weeks ago maybe less

**Comment by ecdye** [comment]
> > The one from months ago, when I first joined? I don't recall making one since.
> 
> No a later one also nominating Holger, something like 6 weeks ago maybe less

Hmm... I don't remember creating a nomination as it goes against openHAB guidelines to self nominate. I'm not opposed to being a maintainer, however I just don't remember creating a nomination. I don't see any post on the team discussion page either, and I don't think I can create posts either, although I could be wrong on that one as I haven't tried.

**Comment by mstormi** [comment]
Ok, never mind, can have been me to be mistaken there. Just wanted to let you know I didn't mean to ignore you there.

**Pull request merged**
